nibabel/tests/test_quaternions.py

```diff
@@ -69,6 +74,53 @@ def test_fillpos():
     assert wxyz[0] == 0.0
 
 
+@pytest.mark.parametrize('dtype', ('f4', 'f8'))
+def test_fillpositive_plus_minus_epsilon(dtype):
+    # Deterministic test for fillpositive threshold
+    # We are trying to fill (x, y, z) with a w such that |(w, x, y, z)| == 1
+    # If |(x, y, z)| is slightly off one, w should still be 0
+    nptype = np.dtype(dtype).type
+
+    # Obviously, |(x, y, z)| == 1
+    baseline = np.array([0, 0, 1], dtype=dtype)
+
+    # Obviously, |(x, y, z)| ~ 1
+    plus = baseline * nptype(1 + np.finfo(dtype).eps)
+    minus = baseline * nptype(1 - np.finfo(dtype).eps)
+
+    assert nq.fillpositive(plus)[0] == 0.0
+    assert nq.fillpositive(minus)[0] == 0.0
+
+
+@pytest.mark.parametrize('dtype', ('f4', 'f8'))
+def test_fillpositive_simulated_error(dtype):
+    # Nondeterministic test for fillpositive threshold
+    # Create random vectors, normalize to unit length, and count on floating point
+    # error to result in magnitudes larger/smaller than one
+    # This is to simulate cases where a unit quaternion with w == 0 would be encoded
+    # as xyz with small error, and we want to recover the w of 0
+
+    # Permit 1 epsilon per value (default, but make explicit here)
+    w2_thresh = 3 * np.finfo(dtype).eps
+
+    pos_error = neg_error = False
+    for _ in range(50):
+        xyz = norm(gen_vec(dtype))
+
+        wxyz = nq.fillpositive(xyz, w2_thresh)
+        assert wxyz[0] == 0.0
+
+        # Verify that we exercise the threshold
+        magnitude = xyz @ xyz
+        if magnitude < 1:
+            pos_error = True
+        elif magnitude > 1:
+            neg_error = True
+
+    assert pos_error, 'Did not encounter a case where 1 - |xyz| > 0'
```

Comment:
But do you still want to fail the tests here, on a random criterion?

Reply:
Good call. Just went ahead and checked the frequencies:

```Python
In [26]: vec = np.array([np.sqrt(np.sum(norm(gen_vec('f4'))**2)) for _ in range(10000)])

In [27]: np.unique(np.sign(1 - vec), return_counts=True)
Out[27]: (array([-1.,  0.,  1.], dtype=float32), array([  81, 6946, 2973]))
```

Seems negative errors are much less likely and we might have failed out for no good reason. I had assumed that the majority would be non-zero error and roughly split in either direction.